frame/transaction-payment/src/lib.rs

```diff
@@ -214,32 +223,61 @@ impl<T: Trait + Send + Sync> SignedExtension for ChargeTransactionPayment<T>
 		info: Self::DispatchInfo,
 		len: usize,
 	) -> TransactionValidity {
-		// pay any fees.
+		let author = T::Author::author();
+
+		let author_initial_balance = T::Currency::total_balance(&author);
+
+		// pay any fees and tip.
 		let tip = self.0;
-		let fee = Self::compute_fee(len as u32, info, tip);
+		// Only mess with balances if tip is not zero.
+		if !tip.is_zero() {
+			let imbalance = match T::Currency::withdraw(
+				who,
+				tip,
+				WithdrawReason::Tip.into(),
+				ExistenceRequirement::KeepAlive,
+			) {
+				Ok(imbalance) => imbalance,
+				Err(_) => return InvalidTransaction::Payment.into(),
+			};
+			T::OnTransactionTipPayment::on_unbalanced(imbalance);
+		}
+
+		let fee = Self::compute_fee(len as u32, info);
 		// Only mess with balances if fee is not zero.
 		if !fee.is_zero() {
 			let imbalance = match T::Currency::withdraw(
 				who,
 				fee,
-				if tip.is_zero() {
-					WithdrawReason::TransactionPayment.into()
-				} else {
-					WithdrawReason::TransactionPayment | WithdrawReason::Tip
-				},
+				WithdrawReason::TransactionPayment.into(),
 				ExistenceRequirement::KeepAlive,
 			) {
 				Ok(imbalance) => imbalance,
 				Err(_) => return InvalidTransaction::Payment.into(),
 			};
-			T::OnTransactionPayment::on_unbalanced(imbalance);
+			T::OnTransactionFeePayment::on_unbalanced(imbalance);
 		}
 
-		let mut r = ValidTransaction::default();
-		// NOTE: we probably want to maximize the _fee (of any type) per weight unit_ here, which
-		// will be a bit more than setting the priority to tip. For now, this is enough.
-		r.priority = fee.saturated_into::<TransactionPriority>();
-		Ok(r)
+		let author_balance_diff = T::Currency::total_balance(&author)
```

Comment:
This section could use a bit of documenting IMO; what is the author doing here? I imagine you are checking if any of the tip/fee was added to the author's balance in the two `if !fee.is_zero() {}` and `if !tip.is_zero() {}` blocks. Can't this be predicted, so that we can prevent another storage access? same goes with accessing `T::Currency::total_issuance()`. Note that this is called per transaction. Anything that we can avoid is a big gain.

Reply:
Indeed it seems that checking author balance is too costy.
I wanted to keep OnTransactionFeePayment generic, but maybe we should require this associated type to return how much is given to author.
And I'll try to remove total_issuance call.

Reply:
So I improved readibility, additional comment about cost:
* author account balance shouldn't be too costy to get, as it is should be in cache as it has been modified anyway so it is would stay in cache.
* getting the author though requires iterating through digest which are empty in-between blocks, and during blocks pallet-authorships save it into a temporary storage.